frame/pallet-xcm-transactor/src/chain_extension.rs

```diff
@@ -0,0 +1,281 @@
+        let origin_location = unwrap!(
+            T::ExecuteXcmOrigin::ensure_origin(origin.into()),
+            InvalidOrigin
+        );
+
+        let hash = input.xcm.using_encoded(sp_io::hashing::blake2_256);
+        // execute XCM
+        // NOTE: not using pallet_xcm::execute here because it does not return XcmError
+        //       which is needed to ensure xcm execution success
+        let outcome = T::XcmExecutor::execute_xcm_in_credit(
+            origin_location,
+            input.xcm.clone(),
+            hash,
+            input.weight,
+            input.weight,
+        );
+
+        // adjust with actual weights used
+        env.adjust_weight(charged, outcome.weight_used());
+        // revert for anything but a complete execution
+        match outcome {
+            Outcome::Complete(_) => Ok(RetVal::Converging(Success.into())),
+            _ => Ok(RetVal::Converging(ExecutionFailed.into())),
+        }
+    }
+
+    /// Returns the fee required to send XCM and saves
+    /// it for sending
+    fn validate_send<E: Ext<T = T>>(
+        &mut self,
+        env: Environment<E, InitState>,
+    ) -> DispatchResult<RetVal> {
+        let mut env = env.buf_in_buf_out();
+        let len = env.in_len();
+        let input: ValidateSendInput = env.read_as_unbounded(len)?;
+
+        let dest = unwrap!(input.dest.try_into(), BadVersion);
+        let xcm: Xcm<()> = unwrap!(input.xcm.try_into(), BadVersion);
+        // validate and get fees required to send
+        let (_, asset) = unwrap!(
+            validate_send::<T::XcmRouter>(dest, xcm.clone()),
+            SendValidateFailed
+        );
+
+        // save the validated input
+        self.validated_send = Some(ValidatedSend { dest, xcm });
+        // write the fees to output
+        VersionedMultiAssets::from(asset).using_encoded(|a| env.write(a, true, None))?;
+
+        Ok(RetVal::Converging(XcmCeError::Success.into()))
+    }
+
+    /// Send the validated XCM
+    fn send<E: Ext<T = T>>(
+        &mut self,
+        mut env: Environment<E, InitState>,
+    ) -> DispatchResult<RetVal> {
+        let input = unwrap!(
+            self.validated_send.as_ref().take().ok_or(()),
+            PreparationMissing
+        );
+
+        let base_weight = <T as pallet_xcm::Config>::WeightInfo::send();
+        env.charge_weight(base_weight)?;
+
+        // TODO: find better way to get origin
+        //       https://github.com/paritytech/substrate/pull/13708
+        let origin = RawOrigin::Signed(env.ext().address().clone());
+
+        // send the xcm
+        unwrap!(
+            XcmPallet::<T>::send(
+                origin.into(),
+                Box::new(input.dest.into()),
+                Box::new(xcm::VersionedXcm::V3(input.xcm.clone())),
+            ),
+            SendFailed
+        );
+
+        Ok(RetVal::Converging(XcmCeError::Success.into()))
+    }
+
+    /// Register the new query
+    /// TODO: figure out weights
```

Comment:
Weight of registration, weight of execution (and payment for it), rent fee (and reimbursment?).

Reply:
Only weight of registration and rent fee (still a TODO).
Execution weight will be paid while sending the XCM calculated via `PriceForSiblingDelivery` in [`cumulus_pallet_xcmp_queue` config which right now is set to `()`](https://github.com/AstarNetwork/Astar/blob/7d2bbebef0b17c5b2bf54a9ba5fde6ed124364cc/runtime/shibuya/src/xcm_config.rs#L254).


This is because the weight of callback/notify is set inside the query instruction itself and query method does not take xcm as input (chicken and egg problem), for example in the below instruction. 
```rust
ReportTransactStatus(QueryResponseInfo {
    destination: (Parent, Parachain(1)).into(),
    query_id,
    max_weight: Weight::from_parts(100_000_000_000_000, 1024 * 1024 * 1024)),    -----> this will be copied to `QueryResponse` that will be sent to querier chain
})
```
For the above query instruction below query response would be sent and `max_weight` is copied to it. The `OnResponse` handler will ensure that notify dispatch's weight is less than `max_weight` before calling it.
```rust
QueryResponse {
	query_id,
	response,
	max_weight: Weight::from_parts(100_000_000_000_000, 1024 * 1024 * 1024)),
	querier: Some(Here),
},
```

Reply:
Ok, I'm not sure that `PriceForSiblingDelivery` is supposed to take care of that.
In the context we're talking about, it's charging the processing of query response (even before it's polled from the smart contract).

I still don't think that current `pallet-xcm` system will be sufficient to handle this scenario properly :slightly_smiling_face: .
BUT - I'd have to spend some time on investigating this to really be sure and make concrete suggestions.